nipype/interfaces/base.py

```diff
@@ -1080,66 +1078,74 @@ def run(self, **inputs):
+                           self.__class__.__name__, mon_fname)
+            mon_sp = ResourceMonitor(proc_pid, freq=mon_freq, fname=mon_fname)
+            mon_sp.start()
+
+        # Grab inputs now, as they should not change during execution
+        inputs = self.inputs.get_traitsfree()
+        outputs = None
         try:
             runtime = self._run_wrapper(runtime)
             outputs = self.aggregate_outputs(runtime)
-            runtime.endTime = dt.isoformat(dt.utcnow())
-            timediff = parseutc(runtime.endTime) - parseutc(runtime.startTime)
-            runtime.duration = (timediff.days * 86400 + timediff.seconds +
-                                timediff.microseconds / 100000.)
-            results = InterfaceResult(interface, runtime,
-                                      inputs=self.inputs.get_traitsfree(),
-                                      outputs=outputs)
-            prov_record = None
-            if str2bool(config.get('execution', 'write_provenance')):
-                prov_record = write_provenance(results)
-            results.provenance = prov_record
         except Exception as e:
+            import traceback
+            # Retrieve the maximum info fast
+            runtime.traceback = traceback.format_exc()
+            # Gather up the exception arguments and append nipype info.
+            exc_args = e.args if getattr(e, 'args') else tuple()
+            exc_args += ('An exception of type %s occurred while running interface %s.' %
+                         (type(e).__name__, self.__class__.__name__), )
+            if config.get('logging', 'interface_level', 'info').lower() == 'debug':
+                exc_args += ('Inputs: %s' % str(self.inputs),)
+
+            runtime.traceback_args = ('\n'.join(['%s' % arg for arg in exc_args]),)
+
+            if force_raise:
+                raise
+        finally:
+            # This needs to be done always
             runtime.endTime = dt.isoformat(dt.utcnow())
             timediff = parseutc(runtime.endTime) - parseutc(runtime.startTime)
             runtime.duration = (timediff.days * 86400 + timediff.seconds +
-                                timediff.microseconds / 100000.)
-            if len(e.args) == 0:
-                e.args = ("")
-
-            message = "\nInterface %s failed to run." % self.__class__.__name__
+                                timediff.microseconds / 1e6)
+            results = InterfaceResult(interface, runtime, inputs=inputs, outputs=outputs,
+                                      provenance=None)
+
+            # Add provenance (if required)
+            if store_provenance:
+                # Provenance will only throw a warning if something went wrong
+                results.provenance = write_provenance(results)
+
+            # Make sure runtime profiler is shut down
+            if enable_rm:
+                import numpy as np
+                mon_sp.stop()
```

Comment:
assuming that this is a blocking call for now.

Reply:
what do you mean?

Reply:
as in it does not return till the process is stopped and all information gathered. i.e. it's not an asynchronous process. i left this comment before i got to resource_profiler. it is indeed a blocking call.

Reply:
yep, it waits for the monitor to shut down.